server/src/main/java/org/apache/druid/discovery/LocalDruidNodeDiscoveryProvider.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.discovery;
+
+import org.apache.druid.server.DruidNode;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BooleanSupplier;
+
+/**
+ * Local node discovery. In this mode, Druid may run a number of
+ * logical servers within a single process, so that node discovery
+ * is all within the same process, but matches a true distributed
+ * node discovery.
+ * <p>
+ * Concurrency is at the level of the entire provider, which is
+ * necessary to coordinate the active state with ongoing node
+ * discovery and listener registrations. Such a "global" lock
+ * is fine because the actions don't occur frequently, nor do the
+ * actions take much time.
+ */
+public class LocalDruidNodeDiscoveryProvider extends DruidNodeDiscoveryProvider implements DruidNodeAnnouncer
+{
+  private class RoleEntry implements DruidNodeDiscovery
+  {
+    @SuppressWarnings("unused")
+    private final NodeRole role;
+    private final Map<DruidNode, DiscoveryDruidNode> nodes = new HashMap<>();
+    private final List<Listener> listeners = new ArrayList<>();
+
+    private RoleEntry(NodeRole role)
+    {
+      this.role = role;
+    }
+
+    private void register(DiscoveryDruidNode node)
+    {
+      List<DruidNodeDiscovery.Listener> targets;
+      synchronized (LocalDruidNodeDiscoveryProvider.this) {
```

Comment:
Seems like `nodes` is not properly synchronized; here the monitor is `LocalDruidNodeDiscoveryProvider.this` but in other methods it's `RoleEntry.this` by way of the method-level `synchronized` keyword.

Reply:
Suggestion: add errorprone's `@GuardedBy` annotation too, which will make the IDE complain at you if something is used without the proper synchronization.